netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -198,6 +295,246 @@ _net_ebpf_sock_addr_update_store_entries()
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_redirect_handle_entry_t, list_entry);
+        if (entry->filter_id == filter_id) {
+            RemoveEntryList(list_entry);
+            ExFreePool(entry);
+            break;
+        }
+        list_entry = list_entry->Flink;
+    }
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+}
+
+static NTSTATUS
+_net_ebpf_ext_sock_addr_get_redirect_handle(uint64_t filter_id, _Out_ HANDLE* redirect_handle)
+{
+    NTSTATUS status = STATUS_NOT_FOUND;
+    KIRQL old_irql;
+
+    old_irql = ExAcquireSpinLockShared(&_net_ebpf_ext_sock_addr_lock);
+
+    LIST_ENTRY* list_entry = _net_ebpf_ext_redirect_handle_list.Flink;
+    while (list_entry != &_net_ebpf_ext_redirect_handle_list) {
+        net_ebpf_extension_redirect_handle_entry_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_redirect_handle_entry_t, list_entry);
+        if (entry->filter_id == filter_id) {
+            *redirect_handle = entry->redirect_handle;
+            status = STATUS_SUCCESS;
+            break;
+        }
+        list_entry = list_entry->Flink;
+    }
+
+    ExReleaseSpinLockShared(&_net_ebpf_ext_sock_addr_lock, old_irql);
+
+    return status;
+}
+
+static void
+_net_ebpf_ext_initialize_connection_context(
+    bool redirected,
+    bool loopback,
+    uint32_t verdict,
+    uint64_t transport_endpoint_handle,
+    uint32_t protocol,
+    _Out_ net_ebpf_extension_connection_context_t* connection_context)
+{
+    connection_context->key.protocol = (uint16_t)protocol;
+    connection_context->value.loopback = loopback;
+    connection_context->value.redirected = redirected;
+    connection_context->value.verdict = (uint8_t)verdict;
+    connection_context->key.transport_endpoint_handle = transport_endpoint_handle;
+    connection_context->timestamp = KeQueryInterruptTime() / 10000; // in ms.
+}
+
+static bool
+_net_ebpf_ext_is_loopback_address(_In_ const bpf_sock_addr_t* address)
+{
+    SOCKADDR_STORAGE socket_address = {0};
+    socket_address.ss_family = (ADDRESS_FAMILY)address->family;
+    if (socket_address.ss_family == AF_INET) {
+        SOCKADDR_IN* v4_address = (SOCKADDR_IN*)&socket_address;
+        v4_address->sin_addr.S_un.S_addr = address->user_ip4;
+    } else {
+        SOCKADDR_IN6* v6_address = (SOCKADDR_IN6*)&socket_address;
+        RtlCopyMemory(v6_address->sin6_addr.u.Byte, address->user_ip6, 16);
+    }
+
+    return INETADDR_ISLOOPBACK((SOCKADDR*)&socket_address);
+}
+
+/**
+ * @brief Compare the destination address in the two provided bpf_sock_addr_t structs.
+ *
+ * @param[in] addr1 Pointer to first sock_addr struct to compare.
+ * @param[in] addr2 Pointer to second sock_addr struct to compare.
+
+ * @return true, if the addresses are same, false otherwise.
+ */
+static inline bool
+_net_ebpf_ext_compare_destination_address(_In_ const bpf_sock_addr_t* addr1, _In_ const bpf_sock_addr_t* addr2)
+{
+    ASSERT(addr1->family == addr2->family);
```

Comment:
Why can't you just memcmp(addr1, addr2, sizeof(*addr1)) ?

(As an aside, such a memcmp would also automatically work once the scope id is part of bpf_sock_addr_t whereas currently the code below arguably has a design flaw by not comparing it)

Reply:
Will add a comment here, but I have also now changed the key saved in the connection context (not using `bpf_sock_addr_t`), and am now using memcpy to find the matching connection context (see `_net_ebpf_ext_get_connection_context()`)

Reply:
Actually this function is now being used only to decide if the eBPF program has modified the destination IP (hence redirected the connection).